Market/Services/PurchaseService.cs

```diff
@@ -0,0 +1,222 @@
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Options;
+
+namespace Market.Services
+{
+    public class PurchaseService : IPurchaseService
+    {
+        private readonly IProductRepository _productRepository;
+        private readonly IPurchaseRepository _purchaseRepository;
+        private readonly ILogger<PurchaseService> _logger;
+        private readonly IConfiguration _configuration;
+        private readonly IMapper _mapper;
+
+        public PurchaseService(IPurchaseRepository purchaseRepository, IProductRepository productRepository, IConfiguration configuration, ILogger<PurchaseService> logger, IMapper mapper)
+        {
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _productRepository = productRepository ?? throw new ArgumentNullException(nameof(productRepository));
+            _purchaseRepository = purchaseRepository ?? throw new ArgumentNullException(nameof(_purchaseRepository));
+            _mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+        }
+
+        public async Task<bool> NewPurchase(List<int> productsId, Purchase purchase)
+        {
+            decimal total = 0;
+            decimal subTotal = 0;
+
+            var products = IsExistProducts(productsId);
+
+            foreach (var product in products)
+            {
+                var purchaseProduct = new PurchaseProducts
+                {
+                    ProductId = product.Id,
+                    Purchase = purchase
+                };
+                subTotal += product.Price;
+                purchase.PurchaseProducts.Add(purchaseProduct);
+            }
+
+            if(purchase.DeliveryType == Purchase.Delivery.Express)
+            {
+                if(string.IsNullOrWhiteSpace(purchase.Address.AdditionalData)) throw new Exception("Additional Data of address is required");
+            }
+            else if(purchase.DeliveryType == Purchase.Delivery.Local)
+            {
+                purchase.Address.AdditionalData = "N/A";
+                purchase.Address.Latitud = 0;
+                purchase.Address.Longitud = 0;
+            }
+
+
+            var impuesto = (double)subTotal * 0.13;
+
+            total = subTotal + (decimal)impuesto;
+
+            purchase.Total = Math.Round(total, 2);
+            purchase.SubTotal = Math.Round(subTotal, 2);
+
+            var result = await _purchaseRepository.AddAsync(purchase);
+            return result;
+        }
+
+        private List<Product> IsExistProducts(List<int> productsId)
+        {
+            List<Product> products = new List<Product>();
+            foreach (var productId in productsId)
+            {
+                Product product = _productRepository.GetById(productId).Result;
+                if (product == null) throw new Exception("Product not exist");
```

Comment:
As in the previous example, it's better to use a specific exception

Reply:
Is better create a exception personalized, because product is not an argument